hudi-spark-datasource/hudi-spark/src/test/scala/org/apache/hudi/functional/TestDagExecutionDataSource.scala

```diff
@@ -0,0 +1,231 @@
+    inputDF.write.format("hudi")
+      .options(commonOpts)
+      .option("hoodie.cleaner.commits.retained", "0")
+      .option("hoodie.parquet.small.file.limit", "0")
+      .option("hoodie.clustering.inline", "true")
+      .option("hoodie.clustering.inline.max.commits", "2")
+      .option("hoodie.metadata.enable", "false")
+      .mode(SaveMode.Append)
+      .save(basePath)
+
+    // verify that clustering is not trigered more than once.
+    assertEquals(sm.triggerCount, 1)
+  }
+
+  /**
+   * Validates that bulk insert dag is triggered only once.
+   * We leverage spark event listener to validate it.
+   */
+  @Test
+  def testValidateBulkInsertForRepeatedDag(): Unit = {
+    // register stage event listeners
+    val sm = new StageEventManager("org.apache.hudi.HoodieSparkSqlWriter$.bulkInsertAsRow")
+    spark.sparkContext.addSparkListener(sm)
+
+    var structType: StructType = null
+    val records = recordsToStrings(dataGen.generateInserts("%05d".format(1), 100)).toList
+    val inputDF = spark.read.json(spark.sparkContext.parallelize(records, 2))
+    structType = inputDF.schema
+    inputDF.write.format("hudi")
+      .options(commonOpts)
+      .option(DataSourceWriteOptions.OPERATION.key, DataSourceWriteOptions.BULK_INSERT_OPERATION_OPT_VAL)
+      .option("hoodie.metadata.enable", "false")
+      .mode(SaveMode.Overwrite)
+      .save(basePath)
+
+    // verify that bulk insert is not trigered more than once.
+    assertEquals(sm.triggerCount, 1)
+  }
+
+  /**
+   * Validates that insert and upsert dag is triggered only once.
+   * We leverage spark event listener to validate it.
+   */
+  @ParameterizedTest
+  @ValueSource(strings = Array("upsert", "insert"))
+  def testValidateInsertAndUpsertForRepeatedDag(operation: String): Unit = {
+    // register stage event listeners
+    val sm = new StageEventManager("org.apache.hudi.client.SparkRDDWriteClient.commit")
+    spark.sparkContext.addSparkListener(sm)
+
+    var structType: StructType = null
+    val records = recordsToStrings(dataGen.generateInserts("%05d".format(1), 100)).toList
+    val inputDF = spark.read.json(spark.sparkContext.parallelize(records, 2))
+    structType = inputDF.schema
+    inputDF.write.format("hudi")
+      .options(commonOpts)
+      .option(DataSourceWriteOptions.OPERATION.key, operation)
+      .option("hoodie.metadata.enable", "false")
+      .mode(SaveMode.Overwrite)
+      .save(basePath)
+
+    // verify that operation is not trigered more than once.
+    assertEquals(sm.triggerCount, 1)
+  }
+
+  /**
+   * Validates that compaction dag is triggered only once.
+   * We leverage spark event listener to validate it.
+   */
+  @Test
+  def testValidateCompactionForRepeatedDag(): Unit = {
+    // register stage event listeners
+    val sm = new StageEventManager("org.apache.hudi.table.action.compact.RunCompactionActionExecutor.execute")
+    spark.sparkContext.addSparkListener(sm)
+
+    var structType: StructType = null
+    for (i <- 1 to 2) {
```

Comment:
why do you need 2 deltacommit here? you just need the first commit then a subsequent update that trigger compaction?